### PR TITLE
Switch view to actual aligned results, not the raw DKIM/SPF results

### PR DESCRIPTION
**Pull request by wolfgangkarall**
This replaces pull request #53

### CODE REVIEW

**Comment by jnew-gh** [comment]
Based on your previous pull request, #53, when I saw this request come in, I was ready to blast you for trying to remove functionality, but I see that this is a much more thoughtful, albeit more complicated, approach to displaying DMARC alignment while retaining the colour-coded visual cues for SPF/DKIM authentication.

However, upon a cursory review of the code, it still seems that some functionality and information display may be negatively affected. I don't have time in the next few days to really go over it in detail, so I would ask @techsneeze not to consider merging this until at least March 17, 2021. I hope to be able to lend some time to this early next week.

**Comment by wolfgangkarall** [comment]
Yes, it's a big change, but one necessary to get the actually important information across: where are DMARC failures for my domains that I need to investigate further. Previously, with the focus on the raw results, that was rather difficult, as those are not really important by themselves as long as the alignment is not taken into account.

**Comment by wolfgangkarall** [comment]
Just so it's not misunderstood: I'm all for getting back some of the information, e.g. filter/drill down regarding SPF/DKIM results (where BTW I agree that it's a loss of info that only one result is stored in the DB by the parser).

And also to filter based on policy_p or disposition, after all there are plenty of 'DMARC failed, but local_policy says it's a forwarder/mailing list we're dealing with' cases.

But it would be nice to have the basics down, e.g. show DMARC results properly (and there is only 'pass' of 'fail' for that).

But looking forward to any feedback. :)

**Comment by jnew-gh** [comment]
I agree.

It's taken me a while to wrap my head around DMARC/SPF/DKIM but it's evident that DMARC alignment is as important as SPF/DKIM authentication when trying to troubleshoot email and adding it to dmarcts will be an improvement.

**Comment by jnew-gh** [comment]
Sorry it's taken me longer than I promised but I wanted to do a thorough review of the changes.

TLDR; I don't recommend merging your code because it negatively impacts functionality but I have used it as a base to include even more functionality.

At this point, I can't recommend accepting this pull request as submitted because there are too many negative impacts, some being
- generally reduces the amount of information conveyed,
- while adding the DKIM/SPF Alignment results, it removes the DKIM/SPF Authorization results,
- the colour code of the reports changes when the DMARC Result filter is used (to be fair, the current code does it too but I think I've found the cause. More on that later.)
- loses the ability to sort the report list by color code,
- those half-circle characters are horrible :-)

**However**, I still think that your goal of showing the DMARC result is necessary, so I have used your code changes and morphed them into something more.

I have set up a public dmarcts server at https://hazelden.ca/dmarcts/dmarcts-report-viewer_new-master-code-on-dmarctest/dmarcts-report-viewer.php populated with sample dmarc data (taken from my dmarc database and somewhat sanitized). Everything should work. 

The Highlights:
- The coloured font circles in the report list have been replaced with rounded divs that form two half circles, left and right.
- The left half-circle shows the DMARC result:
  - lime: all emails passed DMARC
  - orange: some emails passed
  - red: no emails passed
  - yellow: other (is there ever a DMARC result that isn't pass or fail?)
- The right half-circle shows if there are any problems with DKIM/SPF Authorization, Alignment  _and_ DMARC Result (a total of 5 categories)
  - lime: no problems
  - orange: at least one failure in one category
  - red: all failed
  - yellow: other condition (neutral or null result in at least one category)
- The report list is filterable by DMARC Result
  - TODO: Add the ability to filter by the problem codes (the right half-circle)
- The report list is sortable by DMARC Result by clicking on the black left half-circle in the table header.
  - TODO: Add the ability to sort by the problem codes (the right half-circle)

There are some extraneous columns (Min & Max in the report list, DMARC Result in the data table) at this point to help me keep track of the various conditions.

I think I've also identified and fixed the bug where the colour-coded circles would change when using the DMARC Result filter. We have been adding the $where clause to the wrong $sql statement.

I hope everything works. Let me know.

**Comment by wolfgangkarall** [comment]
No having much time today myself, but this looks really nice, and indeed it brings back columns that I skipped for the time being and nicely highlights those 'SPF result pass, but not aligned' cases.

For those I'd actually suggest to use different colors, i.e. lighter green/red for the not so important SPF/DKIM results, but dark green/red for the 'SPF/DKIM Align' columns because that's where the important stuff is for a DMARC viewer.

Can you put this code into some branch in your own repository that I can try it against live data?

**Comment by jnew-gh** [comment]
Thanks for the feedback.

I was thinking of toning down the data report table colours a bit because they are quite harsh. As for different colours for alignment vs. authentication, I wasn't thinking of getting quite that fine-grained. Four colours is already enough.

I was hoping that @techsneeze could chime in before I go too much further to see if we are heading in the right direction. I'll create a pull request on my fork in the next few days.

**Comment by techsneeze** [comment]
First, a big thank you to both of you, for helping to grow this tool and make it more useful for others!  

I do agree that we should probably try to keep the color number minimal, as it can become difficult to visually parse quickly.  Ideally we'd also consider colors that are accessible.  I have a co-worker who is color blind, so this has become important to me.  Once you click into the report, you can probably have more variation there, to provide the nuance of the report status.  

So, it seems as if @jnew-gh has tried to create a "combo" of both of your ideas.  The sample link looks like a good next step.  If you can provide @wolfgangkarall with a view of the code... a pull request or such... then we can probably continue to evolve from there.

**Comment by jnew-gh** [comment]
Thanks @techsneeze for pointing out the accessibility issues. I quickly put the colours we use through this contrast checker (https://webaim.org/resources/contrastchecker/) and all the text/background combinations in the new code meet the WCAG AAA contrast requirements.

However, the colours, when compared to one another, don't. This would mean that we couldn't use these colours as the _sole_ method of differentiating the results. Thankfully, for everything except the colour-coded circles and the overall DMARC Result (in the report data table), there is text that displays the result.

Even before you pointed out the accessibility issue, I was considering using the _title_ attribute to display the non-text colour information when a circle or row is hovered over. Do you think that would suffice? (Although w3.org isn't enthusiastic about it. See https://www.w3.org/WAI/WCAG20/Techniques/ua-notes/html#H33)  I could also add a _DMARC Result_ column to the report data table. I'm not quite sure of an alternate proposal for the circles, though.

As for posting the code, now that you have approved the direction in which this is going, I will try to get to that later this weekend

**Comment by techsneeze** [comment]
Yeah, I think a hover over description would be a good step forward.  I'm sure this will all continue to evolve, so we can wait and think about the other stuff.  It looks like the title tags are supported by some assistive technologies.

**Comment by Maeglin73** [comment]
DKIM/SPF alignment results are indeed simply "pass" and "fail", but dmarcts_report_parser allows for a third possibility ("unknown") for those fields if they aren't actually in the report. The RFC says those fields are supposed to be there, though, so I'd consider "unknown" to be equivalent to "fail" in that case. That's probably something better addressed in the parser script, though, so whether or not to account for it here is up to you guys.

**Comment by wolfgangkarall** [comment]
@Maeglin73 I was thinking about merging 'unknown' into 'fail' when working on the parser, but I think it's best to leave the information available that a reporting organization didn't provide an actual fail, e.g. in case the actual XML reports are deleted. But I'm all for interpreting the 'unknown' as 'fail' when it comes to the viewer.

**Comment by Maeglin73** [comment]
The spec only allows for "pass" and "fail" in those fields, and the parser is the one that sets it as "unknown" if the alignment result fields are not present. The only time I've seen one of them not present is if the DKIM validation results were missing entirely, as in there was no DKIM signature to verify. I'd call a missing DKIM signature the result of either spoofing, or a half-assed DMARC config.

If you don't want to always make them equivalent in the code, then perhaps make it an option?

**Comment by jnew-gh** [comment]
> I'd call a missing DKIM signature the result of either spoofing, or a half-assed DMARC config.

Exactly why it should remain flagged as a problem. I myself have implemented half-assed DMARC configs. :-)

> If you don't want to always make them equivalent in the code, then perhaps make it an option?

I'm currently writing an options screen for the viewer, and I have put in that option as a placeholder so I don't forget about it.

**Comment by techsneeze** [comment]
I agree that ensuring there is space for unknown is important, so it can be investigated.  We've seen a LOT of poorly implemented DMARC reporting systems, that don't quite follow the specification.  While we could just ignore them as being badly formatted, we'd miss out on a lot of important reports.  

With the other work that's been going on, this pull has deviated from master quite a bit and has conflicts.  is @jnew-gh or @wolfgangkarall still working on another approach?

**Comment by jnew-gh** [comment]
I submitted a pull request to @wolfgangkarall's repository a couple of weeks ago. We had some discussions then but nothing for awhile now. The pull requests I've submitted to master here shouldn't interfere too much with the new code. If you like, I could submit my take on the alignment presentation in the next day or so. I'll take some of wolfgangkarall's suggestions and integrate them with my code.

**Comment by jnew-gh** [comment]
We can close this now as #62 implements the functionality of this pull request.

**Comment by techsneeze** [comment]
Closing per your request.
